ext/test/key_provider/key_provider.c

```diff
@@ -0,0 +1,404 @@
+        kp->state.current_lsn = crypt->r.lsn;
+
+        assert(memcmp(kp->state.current_key, crypt->keys.data, kp->state.key_size) == 0);
+        assert(kp->state.key_size == crypt->keys.size);
+    }
+
+    return (0);
+}
+
+/*
+ * kp_terminate --
+ *     Cleanup function called when the key provider is being shut down.
+ */
+static int
+kp_terminate(WT_KEY_PROVIDER *wtkp, WT_SESSION *session)
+{
+    KEY_PROVIDER *kp = (KEY_PROVIDER *)wtkp;
+
+    LOG_INFO(kp, session, "Terminating key provider");
+
+    if (kp->state.current_key != NULL)
+        free(kp->state.current_key);
+
+    free(kp);
+
+    return (0);
+}
+
+#define CONFIGURE_BEGIN(kp) \
+    if (kp == NULL) {       \
+        assert(kp != NULL); \
+        return (EINVAL);    \
+    }
+
+#define CONFIGURE_PARAM(kp, param, k, v, ctype, wt_type)                                   \
+    else if (strncmp(#param, k.str, k.len) == 0 && k.len == strlen(#param) && v.len > 0 && \
+      v.type == wt_type)                                                                   \
+    {                                                                                      \
+        kp->param = (ctype)v.val;                                                          \
+        continue;                                                                          \
+    }
+
+#define CONFIGURE_INT(kp, param, k, v) CONFIGURE_PARAM(kp, param, k, v, int, WT_CONFIG_ITEM_NUM)
+#define CONFIGURE_UINT(kp, param, k, v) \
+    CONFIGURE_PARAM(kp, param, k, v, unsigned int, WT_CONFIG_ITEM_NUM)
+
+#define CONFIGURE_END(kp, k, v)                                                           \
+    else                                                                                  \
+    {                                                                                     \
+        LOG_ERROR(kp, NULL, "WT_CONFIG_PARSER.next: unexpected configuration: %.*s=%.*s", \
+          (int)k.len, k.str, (int)v.len, v.str);                                          \
+        ret = EINVAL;                                                                     \
+        goto err;                                                                         \
+    }
+
+/*
+ * kp_configure --
+ *     Parse configuration options for the key provider.
+ */
+static int
+kp_configure(KEY_PROVIDER *kp, WT_CONFIG_ARG *config)
+{
+    WT_EXTENSION_API *wtext = kp->wtext;
+    WT_CONFIG_PARSER *config_parser = NULL;
+    WT_CONFIG_ITEM k = {0}, v = {0};
+    int ret = 0;
+
+    if ((ret = wtext->config_parser_open_arg(wtext, NULL, config, &config_parser)) != 0) {
+        LOG_ERROR(kp, NULL, "WT_EXTENSION_API.config_parser_open_arg: error: %d (%s)", ret,
+          wtext->strerror(wtext, NULL, ret));
+        goto err;
+    }
+
+    /* Parse configuration key-value pairs */
+    while ((ret = config_parser->next(config_parser, &k, &v)) == 0) {
+        CONFIGURE_BEGIN(kp)
+        CONFIGURE_INT(kp, verbose, k, v)
```

Comment:
I'm not sure if this reads easier, the fail_fs.c file has a different approach to how config parsing works. e.g.

```
        if (fail_fs_arg("environment", &k, &v, &argval)) {
            fail_fs->use_environment = (argval != 0);
            continue;
        } else if (fail_fs_arg("verbose", &k, &v, &argval)) {
            fail_fs->verbose = (argval != 0);
            continue;
        } else if (fail_fs_arg("allow_writes", &k, &v, &fail_fs->allow_writes))
            continue;
        else if (fail_fs_arg("allow_reads", &k, &v, &fail_fs->allow_reads))
 ```